File: app/src/main/java/com/example/narthana/genpass/Utility.java
package com.example.narthana.genpass;

import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;
import android.util.TypedValue;

import com.example.narthana.genpass.WordContract.WordEntry;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

/**
 * Created by narthana on 22/10/16.
 */

class Utility
{
    private static Random r;

    static void loadDictionary(Context context, InputStream dictionary)
    {
        try (BufferedReader br = new BufferedReader(new InputStreamReader(dictionary));
             SQLiteDatabase db = new NewWordDBHelper(context).getWritableDatabase())
        {
            db.beginTransaction();
            db.delete(WordEntry.TABLE_NAME, null, null);
            ContentValues content = new ContentValues(2);
            for (String word; (word = br.readLine()) != null; )
            {
                content.clear();
                content.put(WordEntry.COLUMN_WORD, word);
                content.put(WordEntry.COLUMN_LEN, word.length());
                db.insert(WordEntry.TABLE_NAME, null, content);
            }
            db.setTransactionSuccessful();
            db.endTransaction();
        }
        catch (IOException e)
        {
            Log.e(context.getClass().getName(), "Could not read text file");
            e.printStackTrace();
        }
    }

    static int[] compressWithRanges(int[] input)
    {
        List<Integer> ranges = new ArrayList<>(input.length / 4);
        ranges.add(input.length);
        for (int i = 0, n = input.length - 1; i <= n; ++i)
        {
            int j = i;
            while (i < n && input[i + 1] == input[i] + 1) ++i;
            if (i != j)
            {
                ranges.add(input[j]);
                ranges.add(input[i]);
            }
            else ranges.add(-input[i]);
        }
        int[] out = new int[ranges.size()];
        for (int i = 0; i < out.length; ++i) out[i] = ranges.get(i);
        return out;
    }

    static int[] expandFromRanges(int[] input)
    {
        int[] out = new int[input[0]];
        for (int i = 0, j = 1; j < input.length; )
        {
            if (input[j] < 0)
            {
                out[i++] = -input[j++];
            }
            else
            {
                int n = input[j + 1] - input[j] + 1;
                for (int k = 0; k < n; ++k, ++i) out[i] = input[j] + k;
                j += 2;
            }
        }
        return out;
    }

    static void shuffle(char[] array)
    {
        if (r == null) r = new Random();
        for (int i = array.length - 1; i > 0; --i) swapChar(array, i, r.nextInt(i + 1));
    }

    // shuffle first n elements of the array
    static void shuffleN(int[] array, int n)
    {
        if (r == null) r = new Random();
        for (int i = 0; i < n; ++i)
        {
            int j = r.nextInt(array.length - i) + i; // random int in [i, words.length)
            swapInt(array, i, j);
        }
    }

    private static void swapChar(char[] array, int i, int j)
    {
        if (i != j)
        {
            char temp = array[i];
            array[i] = array[j];
            array[j] = temp;
        }
    }

    private static void swapInt(int[] array, int i, int j)
    {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    static int dpToPx(Context context, int length)
    {
        return (int) TypedValue.applyDimension(
                TypedValue.COMPLEX_UNIT_DIP,
                length,
                context.getResources().getDisplayMetrics()
        );
    }
}

File: app/src/main/java/com/example/narthana/genpass/Utility.kt
package com.example.narthana.genpass

import android.app.Fragment
import android.content.ContentValues
import android.content.Context
import android.preference.PreferenceManager
import com.example.narthana.genpass.data.NewWordDBHelper
import com.example.narthana.genpass.data.WordContract
import java.io.InputStream
import java.util.*

/**
 * Created by narthana on 22/10/16.
 */

fun loadDictionary(context: Context, dictionary: InputStream) {
    NewWordDBHelper(context).writableDatabase.use { db ->
        db.beginTransaction()
        db.delete(WordContract.WordEntry.TABLE_NAME, null, null)
        val content = ContentValues(2)
        dictionary.bufferedReader().useLines { lines ->
            lines.forEach { line ->
                content.run {
                    clear()
                    put(WordContract.WordEntry.COLUMN_WORD, line)
                    put(WordContract.WordEntry.COLUMN_WORD, line.length)
                    db.insert(WordContract.WordEntry.TABLE_NAME, null, this)
                }
            }
        }
        db.setTransactionSuccessful()
        db.endTransaction()
    }
}

fun compressWithRanges(input: IntArray): IntArray = input.foldIndexed(mutableListOf<Int>()) {
    i, list, x -> list.apply {
        if (i == 0) add(if (x + 1 == input[1]) x else -x)
        else if (i == input.lastIndex) add(if (x - 1 == input[i - 1]) x else -x)
        else if (x - 1 != input[i - 1] && x + 1 != input[i + 1]) add(-x)
        else if (x - 1 != input[i - 1] || x + 1 != input[i + 1]) add(x)
    }
}.toIntArray()

fun expandFromRanges(input: IntArray): IntArray {
    val (singles, ranges) = input.partition { it < 0 }
    val (starts, ends) = ranges.partitionIndexed { i, _ -> i % 2 == 0 }
    val expandedRanges = starts.zip(ends).flatMap { IntRange(it.first, it.second).toList() }
    return (singles.map { -it } + expandedRanges).toIntArray()
}

fun<T> Iterable<T>.partitionIndexed(predicate: (index: Int, T) -> Boolean): Pair<List<T>, List<T>> {
    val first = mutableListOf<T>()
    val second = mutableListOf<T>()
    var i = 0
    for (x in this) if (predicate(i++, x)) first.add(x) else second.add(x)
    return Pair(first, second)
}

fun shuffle(array: CharArray, r: Random) {
    fun swapChar(i: Int, j: Int) {
        if (i != j) {
            val temp = array[i]
            array[i] = array[j]
            array[j] = temp
        }
    }

    for (i in array.lastIndex downTo 1) swapChar(i, r.nextInt(i + 1))
}

// shuffle first n elements of the array
fun shuffleFirst(array: IntArray, n: Int, r: Random) {
    fun swapInt(i: Int, j: Int) {
        if (i != j) {
            val temp = array[i]
            array[i] = array[j]
            array[j] = temp
        }
    }

    for (i in 0 until n) swapInt(i, r.nextInt(array.lastIndex) + i)
}

/**
 * Selects [m] random elements from [array]. The selection uniform.
 *
 * @param array the array from which to obtain random elements
 * @m the number of randoms, must be in the range 1..[array.size]
 * @r an instance of [Random]
 * @return an array containing the random elements
 */
fun randomN(array: IntArray, m: Int, r: Random): IntArray {
    var ranges = linkedListOf(IntRange(0, array.lastIndex))
    val nums = IntArray(m) { 0 }
    for (i in 0 until m) {
        val temp = randFromRanges(ranges, r)
        ranges = temp.first
        nums[i] = array[temp.second]
    }
    return nums
}

val IntRange.len: Int
    get() = last - first + 1

/**
 * Takes a mutable list of ranges and picks a int uniformly at random from the ranges.
 * Splits the range at the chosen int and returns a Pair of the new list and the int
 *
 * @param ranges a list of [IntRange] from which to select a random integer
 * @param r an instance of [Random]
 * @return A [Pair]
 */
fun randFromRanges(ranges: LinkedList<IntRange>, r: Random): Pair<LinkedList<IntRange>, Int> {
    var hole = -1
    var rangeList = ranges.filter { !it.isEmpty() }
    val selection = rangeList.map { it.len }.sum()

    fun rec(p: LinkedList<IntRange>, n: Int): LinkedList<IntRange> = when (p) {
        is EmptyLinkedList -> p
        is NonEmptyLinkedList<IntRange> ->
            if (n < p.head.len) {
                hole = p.head.start + n
                linkedListOf(
                        IntRange(p.head.start, hole - 1),
                        IntRange(hole + 1, p.head.endInclusive)
                ) + p.tail
            }
            else p.apply { tail = rec(tail, n - p.head.len) }
    }

    rangeList = rec(rangeList, selection)
    return Pair(rangeList, hole)
}

sealed class WordListResult
data class WordList(val array: IntArray, val minWordLen: Int, val maxWordLen: Int):
        WordListResult() {
    override fun equals(other: Any?): Boolean = other === this
            && other is WordList
            && Arrays.equals(other.array, array)
            && other.minWordLen == minWordLen
            && other.maxWordLen == maxWordLen
    override fun hashCode(): Int {
        var code = Arrays.hashCode(array)
        code += 31 * code + minWordLen
        code += 31 * code + maxWordLen
        return code
    }
}
object WordListError: WordListResult()

fun Fragment.getStringSet(key: String, defValues: Set<String>?): Set<String>? =
        PreferenceManager.getDefaultSharedPreferences(this.activity).getStringSet(key, defValues)
fun Fragment.getInt(key: String, defValues: Int): Int =
        PreferenceManager.getDefaultSharedPreferences(this.activity).getInt(key, defValues)
fun Fragment.getStringPref(key: String, defValues: String): String =
        PreferenceManager.getDefaultSharedPreferences(this.activity).getString(key, defValues)
fun Fragment.getBoolean(key: String, defValues: Boolean): Boolean =
        PreferenceManager.getDefaultSharedPreferences(this.activity).getBoolean(key, defValues)

